### DISCUSSION TITLE
BigQueryInsertJobOperator Usage

### DISCUSSION THREAD

**Discussion by andresitorusz** [apache/airflow#34140] [Q&A] [answered]
Hi everyone,

Today I'm trying to use BigQueryInsertJobOperator to insert a query job (merge query). Both source and destination table has partitioning and the partition filter is required.

Here are both tables' details:
![Screenshot 2023-09-07 003920](https://github.com/apache/airflow/assets/36465150/18b162b8-ee06-467e-9708-326839d05c64)
![Screenshot 2023-09-07 003946](https://github.com/apache/airflow/assets/36465150/ea69beda-983d-4bc2-8414-16dc748b2b58)

Here is the BigQueryInsertJobOperator hardcoded configuration:

```
query = """DECLARE
  formatted_dates ARRAY<DATE>;
EXECUTE IMMEDIATE
  (
  SELECT
    CONCAT('SELECT [', ARRAY_TO_STRING(ARRAY(
        SELECT
          CONCAT("DATE '", FORMAT_DATE('%Y-%m-%d', flattened_dates_array), "'")
        FROM (
          SELECT
            DISTINCT DATE(created_at) AS flattened_dates_array
          FROM
            `pocz-389704.foo.bar_temp` ) AS t
        ORDER BY
          flattened_dates_array ), ', '), '] AS formatted_dates') );
MERGE
  `pocz-389704.foo.bar` AS x
USING
  `pocz-389704.foo.bar_temp` AS y
ON
  COALESCE(CAST(x.`id` AS string), 'NULL') = COALESCE(CAST(y.`id` AS string), 'NULL')
  AND DATE(x.created_at) IN UNNEST(formatted_dates)
  WHEN MATCHED THEN DELETE
  WHEN NOT MATCHED
  THEN
INSERT
  (`id`,
    `name`,
    `created_at`,
    `updated_at`,
    `load_timestamp`)
VALUES
  (`id`, `name`, `created_at`, `updated_at`, `load_timestamp`)"""

time_partitioning = {
    "type": "DAY",
    "field": self.target_bq_partition_key
}

configuration = {
    "query": {
        "query": query,
        "useLegacySql": False,
        "timePartitioning": time_partitioning
    }
}
```

Then, I got this error: `If partitioning specification is present, destination table must also be present.`

Here is the full log of the task: [dag_id=foo.bar_run_id=scheduled__2023-09-05T17_00_00+00_00_task_id=merge__bar_attempt=3.log](https://github.com/apache/airflow/files/12541015/dag_id.foo.bar_run_id.scheduled__2023-09-05T17_00_00%2B00_00_task_id.merge__bar_attempt.3.log)

I also tried to use this config (but still got the same error):
```
configuration = {
    "query": {
        "query": query,
        "useLegacySql": False,
        "timePartitioning": time_partitioning
    },
    "destinationTable": {
        "datasetId": "foo",
        "projectId": "pocz-389704",
        "tableId": "bar"
    },
}
```

Please help me to fix this. Thanks in advance.

**Comment by Taragolis** [comment]
I'm not familiar with BigQuery, but did you try to put `destinationTable` inside of `query`?

https://github.com/apache/airflow/blob/b9acffa81bf61dcf0c5553942c52629c7f75ebe2/tests/system/providers/google/cloud/bigquery/example_bigquery_queries.py#L139-L153

**Reply by andresitorusz** [reply]
Thank you, I just realized that the destinationTable is misplaced. I just tried to rerun the task using this configuration:
```
configuration = {
    "query": {
        "query": query,
        "useLegacySql": False,
        "timePartitioning": time_partitioning,
        "destinationTable": {
            "datasetId": "foo",
            "projectId": "pocz-389704",
            "tableId": "bar"
        },
    }
}
```
Then, I got this error: `400 configuration.query.destinationTable cannot be set for scripts`

**Comment by andresitorusz** [comment, marked as answer]
The problem I am facing is caused by incorrect configuration. Because in this case, I am inserting a merge query, I should not need to fill in the timePartitioning field in the configuration.

From what I understand, it seems that timePartitioning is intended for job queries with destinationTable. So, we don't need it for the merge query job.
